decidim-proposals/app/permissions/decidim/proposals/permissions.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Proposals
+    class Permissions
+      def self.allows?(user, action, context)
+        new(user, action, context).allowed?
+      end
+
+      def initialize(user, action, context)
+        @user = user
+        @action = action
+        @context = context
+      end
+
+      def allowed?
+        # Stop checks if the user is not authorized to perform the
+        # action for this space
+        return false unless spaces_allows_user?
+        return false unless user
+
+        # Delegate the admin permission checks to the admin permissions class
+        return Decidim::Proposals::Admin::Permissions.allows?(user, action, context) if action[:scope] == "admin"
+        return false if action[:scope] != "public"
+
+        return false if action[:subject] != "proposal"
+
+        return true if can_create_proposal?
```

Comment:
This implementation is a bit misleading, as it looks like you're checking each permission without a context. I'd expect a `case`:

```ruby
case action[:action]
  when :create
    can_create_proposal? # You can remove the action check from the method implementation
  when :edit
    can_edit_proposal?
  else
    false
end
```

Reply:
I wanted to reduce complexity from the `allowed?` method, I feel my implementation is easier to reason about, but I don't have a strong opinion...

Reply:
As a first reader, I thought that we were returning `true` if *in any case* a user can create a proposal. Using a case makes it explicitly that we're branching, and not sequentially executing it. I don't have a strong opinion, but it reminded me a bit of redux and its reducers, which tackle a similar problem (dispatching an action) in a similar manner.

Reply:
Example: https://redux.js.org/recipes/structuring-reducers/refactoring-reducers-example